plugins/inventory/cloudstack.py

```diff
@@ -0,0 +1,288 @@
+EXAMPLES = '''
+# inventory_cloudstack.yml file in YAML format
+# Example command line: ansible-inventory --list -i inventory_cloudstack.yml
+plugin: ngine_io.cloudstack.cloudstack
+
+# Use the default ip as ansible_host
+hostname: v4_default_ip
+
+# Return only instances related to the VPC vpc1 and in the zone EU
+filter_by_vpc: vpc1
+filter_by_zone: EU
+
+# Group instances with a disk_offering as storage
+# Create a group dmz for instances connected to the dmz network
+groups:
+  storage: disk_offering is defined
+  dmz: "'dmz' in networks"
+
+# Group the instances by network, with net_network1 as name of the groups
+# Group the instanes by custom tag sla, groups like sla_value for tag sla
+keyed_groups:
+  - prefix: net
+    key: networks
+  - prefix: sla
+    key: tags.sla
+
+
+'''
+
+# The J2 Template takes 'instance' object as returned from ACS and returns 'instance' object as returned by
+# This inventory plugin.
+# The data structure of this inventory has been designed according to the following criteria:
+# - do not duplicate/compete with Ansible instance facts
+# - do not duplicate/compete with Cloudstack facts modules
+# - hide internal ACS structures and identifiers
+# - if possible use similar naming to previous inventory script
+# - prefer non-existing attributes over null values
+# - populate the data required to group and filter instances
+INVENTORY_NORMALIZATION_J2 = '''
+---
+instance:
+
+  name: {{instance.name}}
+  hostname: {{instance.hostname or instance.name | lower }}
+  v4_default_ip: {{instance.nic[0].ipaddress}}
+
+  zone: {{instance.zonename}}
+  domain: {{instance.domain | lower}}
+  account: {{instance.account}}
+  username: {{instance.username}}
+  {% if instance.group %}
+  group: {{instance.group}}
+  {% endif %}
+
+  {% if instance.tags %}
+  tags:
+  {% for tag in instance.tags %}
+    {{tag.key}}: {{tag.value}}
+  {% endfor %}
+  {% endif %}
+
+  template: {{instance.templatename}}
+  service_offering: {{instance.serviceofferingname}}
+  {% if instance.diskofferingname is defined %}
+  disk_offering: {{instance.diskofferingname}}
+  {% endif %}
+  {% if instance.affinitygroup %}
+  affinity_groups:
+    {% for ag in instance.affinitygroup %}
+    - {{ ag.name }}
+    {% endfor %}
+  {% endif %}
+  networks:
+    {% for nic in instance.nic %}
+    - {{nic.networkname}}
+    {% endfor %}
+
+  ha_enabled: {{instance.haenable}}
+  password_enabled: {{instance.passwordenabled}}
+
+  hypervisor: {{instance.hypervisor | lower}}
+  cpu_speed: {{instance.cpuspeed}}
+  cpu_number: {{instance.cpunumber}}
+  memory: {{instance.memory}}
+  dynamically_scalable: {{instance.isdynamicallyscalable}}
+
+  state: {{instance.state}}
+  cpu_usage: {{instance.cpuused}}
+  created: {{instance.created}}
+'''
+
+from ansible.plugins.inventory import BaseInventoryPlugin, Constructable, AnsibleError
+from ansible.module_utils.basic import missing_required_lib
+from ..module_utils.cloudstack import HAS_LIB_CS, cs_get_api_config
```

Comment:
I am not sure if this works in every case (module not executed locally?) Let me verify

Reply:
should be fine, never mine.

Reply:
Yes, as far as I understand the plugin is not a module.
Here lies the main reason why, if we want to re-use code we need to re-architect.
It would be desirable to share: API logic, Connection logic, document fragments, etc. But he plugin class does not inherit from a module base class where the functionality has been placed.